README.md
## Function e-Bill

> 信用卡交易记录邮件解析

## 招商银行

- 每日信用管家（新）
- 信用管家消费提醒（旧）


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres
to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.3] - 2021-11-15

### Bugfix

- PDF 解析库更换，修复账单文件升级的解析问题
- 账单文件升级导致的字段对应问题修复

## [1.2.2] - 2021-09-02

### Bugfix

- 修复退款记录解析问题

## [1.2.1] - 2021-07-02

### Added

- 使用 Coding 持续集成服务部署 

## [1.2.0] - 2021-07-02

### Bugfix

- 修复 2021-05-26 之后的邮件解析问题

## [1.1.1] - 2020-11-18

### Bugfix

- 上传文件临时写入/tmp, 云函数仅该目录有访问权限
- 修复异常文件未删除的问题

## [1.1.0] - 2020-11-18

### Added

- 解析招行pdf月账单

## [1.0.2] - 2020-11-12

### Bugfix

- 修复超过三位数金额解析问题

## [1.0.1] - 2020-09-22

### Bugfix

- Lower case field in result json

## [1.0.0] - 2020-09-22

### Added

- Parse cmb email
- Automatically build and deploy with github workflow depend on [funcraft](https://github.com/alibaba/funcraft)
- Automatically publish github release with [goreleaser](https://goreleaser.com/)

main.go
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"syscall"
	"time"

	"github.com/gin-gonic/gin"
)

type fetchConfig struct {
	Username string `form:"username"`
	Password string `form:"password"`
	Hour     int64  `form:"hour"`
}

func main() {
	r := gin.New()
	r.Use(gin.Recovery())

	setupRouter(r)

	start(&http.Server{
		Addr:    fmt.Sprintf(":%s", getenv("FC_SERVER_PORT", "9000")),
		Handler: r,
	})
}

func setupRouter(r *gin.Engine) {
	r.GET("/ping", func(c *gin.Context) {
		c.Data(http.StatusOK, "text/plain; charset=utf-8", []byte("pong"))
	})
	r.POST("/cmb", func(c *gin.Context) {
		var cfg fetchConfig
		if c.ShouldBindJSON(&cfg) != nil {
			c.JSON(http.StatusOK, failed("missing required body"))
			return
		}
		if cfg.Hour == 0 {
			cfg.Hour = 24
		}

		orders, err := emailParseCmb(cfg)
		if err != nil {
			c.JSON(http.StatusOK, failed(err.Error()))
			return
		}

		c.JSON(http.StatusOK, data(orders))
	})

	r.POST("/file/cmb", func(c *gin.Context) {
		file, err := c.FormFile("file")
		if err != nil {
			c.JSON(http.StatusOK, failed(err.Error()))
			return
		}
		filename := "/tmp/" + filepath.Base(file.Filename)
		if err := c.SaveUploadedFile(file, filename); err != nil {
			c.JSON(http.StatusOK, failed(err.Error()))
			return
		}
		defer func() {
			_ = os.Remove(filename)
		}()

		orders, err := fileParseCmb(filename)
		if err != nil {
			c.JSON(http.StatusOK, failed(err.Error()))
			return
		}

		c.JSON(http.StatusOK, data(orders))
	})
}

func start(srv *http.Server) {
	go func() {
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Printf("listen: %s\n", err)
		}
	}()

	log.Printf("Start Server @ %s", srv.Addr)
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit
	log.Print("Shutdown Server ...")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := srv.Shutdown(ctx); err != nil {
		log.Printf("Server Shutdown:%s", err)
	}
	<-ctx.Done()
	log.Print("Server exiting")
}

func getenv(key, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}
	return fallback
}

func failed(msg string) gin.H {
	return gin.H{
		"msg":       msg,
		"timestamp": time.Now().Unix(),
	}
}

func data(data interface{}) gin.H {
	return gin.H{
		"msg":       "success",
		"data":      data,
		"timestamp": time.Now().Unix(),
	}
}


mailparse.go
package main

import (
	"github.com/PuerkitoBio/goquery"
	"regexp"
	"strings"
	"time"
)

func parse(doc *goquery.Document) []emailBillOrder {
	var orders = make([]emailBillOrder, 0)
	var orderList []string
	dateStr := time.Now().Add(-24 * time.Hour).Format("2006-01-02")
	doc.Find("#fixBand3 > table > tbody > tr").Each(func(i int, s *goquery.Selection) {
		text := regexp.MustCompile(`\s+`).ReplaceAllString(s.Text(), "")
		if i > 0 {
			// 11:26:34CNY 12.00尾号3885 消费 支付宝-南京天使谷酒店管理有限公司
			orderList = append(orderList, text)
		} else {
			dateStr = strings.TrimSpace(text[:strings.Index(text, "您的消费")])
		}
	})
	dateStr = strings.ReplaceAll(dateStr, "/", "-")
	for _, order := range orderList {
		// 获取交易时间/金额/描述
		for _, match := range orderCNYRe.FindAllStringSubmatch(order, -1) {
			timeStr := strings.TrimSpace(match[1])
			amountStr := strings.TrimSpace(strings.ReplaceAll(match[2], ",", ""))
			name := strings.TrimSpace(match[3])

			orders = append(orders, emailBillOrder{Name: name, Time: dateStr + " " + timeStr, Amount: amountStr})
		}
	}
	return orders
}

func parseBefore20210526(text string) []emailBillOrder {
	var orders = make([]emailBillOrder, 0)
	var orderList []string
	idx := 0
	// 获取所有交易明细 -> 07:53:41人民币 8.00尾号3885 消费 支付宝-暖煨堂餐饮管理有限公司
	for i, match := range timeRe.FindAllString(text, -1) {
		if i == 0 {
			idx = strings.Index(text, match)
			continue
		}
		tmpIdx := strings.Index(text, match)
		orderList = append(orderList, strings.TrimSpace(text[idx:tmpIdx]))
		idx = tmpIdx
	}
	orderList = append(orderList, strings.TrimSpace(text[idx:strings.Index(text, "人民币消费")]))

	// 获取消费日期
	// 邮件内容变更兼容旧邮件
	dateStr := time.Now().Add(-24 * time.Hour).Format("2006-01-02")
	if strings.Index(text, "截至") != -1 && strings.Index(text, "24时") != -1 {
		dateStr = strings.TrimSpace(text[strings.Index(text, "截至")+len("截至") : strings.Index(text, "24时")])
	} else if strings.Index(text, "消费人民币") != -1 {
		dateStr = strings.TrimSpace(text[strings.Index(text, "消费人民币")-12 : strings.Index(text, "消费人民币")-2])
	}
	dateStr = strings.ReplaceAll(dateStr, "/", "-")

	for _, order := range orderList {
		// 获取交易时间/金额/描述
		for _, match := range orderRe.FindAllStringSubmatch(order, -1) {
			timeStr := strings.TrimSpace(match[1])
			amountStr := strings.TrimSpace(strings.ReplaceAll(match[2], ",", ""))
			name := strings.TrimSpace(match[3])

			orders = append(orders, emailBillOrder{Name: name, Time: dateStr + " " + timeStr, Amount: amountStr})
		}
	}
	return orders
}


cmb.go
package main

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"github.com/NoF0rte/pdf"
	"io"
	"io/ioutil"
	"log"
	"math"
	"mime"
	"mime/multipart"
	"net/mail"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
	"github.com/emersion/go-imap"
	"github.com/emersion/go-imap/client"
	"golang.org/x/text/encoding/simplifiedchinese"
)

// emailBillOrder 邮件订单
type emailBillOrder struct {
	Name   string `json:"name"`
	Time   string `json:"time"`
	Amount string `json:"amount"`
}

// fileBillOrder 出账单订单, 交易日,记账日,交易摘要,人民币金额,卡号末四位,交易地金额
type fileBillOrder struct {
	TradingDate   string `json:"tradingDate"`
	CreditDate    string `json:"creditDate"`
	Name          string `json:"name"`
	Amount        string `json:"amount"`
	TailNumber    string `json:"tailNumber"`
	TradingAmount string `json:"tradingAmount"`
}

var (
	// 07:53:41
	timeRe = regexp.MustCompile(`(\d{2}:\d{2}:\d{2})`)
	// 07:53:41人民币 8.00尾号3885 消费 支付宝-xxxx有限公司
	orderRe    = regexp.MustCompile(`(?m)([\d:]+)人民币([  \-\d.,]+)(.*)`)
	orderCNYRe = regexp.MustCompile(`(?m)([\d:]+)CNY([  \-\d.,]+)(.*)`)
	// 2021-05-26
	date20210526 = time.Date(2021, 5, 26, 0, 0, 0, 0, time.Local)
	// 2021-10-31
	date20211031 = time.Date(2021, 10, 31, 0, 0, 0, 0, time.Local)
)

func emailParseCmb(cfg fetchConfig) ([]emailBillOrder, error) {
	var orders = make([]emailBillOrder, 0)
	c, err := client.DialTLS("imap.exmail.qq.com:993", nil)
	if err != nil {
		return orders, fmt.Errorf("dial imap server error: %s", err.Error())
	}
	defer func() {
		_ = c.Logout()
	}()

	if err := c.Login(cfg.Username, cfg.Password); err != nil {
		return orders, fmt.Errorf("email login error: %s", err.Error())
	}

	_, err = c.Select("inbox", true)
	if err != nil {
		return orders, fmt.Errorf("open inbox error: %s", err.Error())
	}

	filter := imap.NewSearchCriteria()
	filter.Since = time.Now().Add(time.Duration(-cfg.Hour) * time.Hour)
	seqNums, err := c.Search(filter)
	if err != nil {
		return orders, fmt.Errorf("search inbox error: %s", err.Error())
	}
	if len(seqNums) == 0 {
		log.Println("no new message")
		return orders, nil
	}

	seqSet := new(imap.SeqSet)
	for _, seqNum := range seqNums {
		seqSet.AddNum(seqNum)
	}

	messages := make(chan *imap.Message, 10)
	done := make(chan error, 1)
	section := &imap.BodySectionName{}
	go func() {
		done <- c.Fetch(seqSet, []imap.FetchItem{section.FetchItem()}, messages)
	}()

	for msg := range messages {
		r := msg.GetBody(section)
		if r == nil {
			log.Println("Server didn't returned message body")
			continue
		}
		mr, err := mail.ReadMessage(r)
		if err != nil {
			log.Printf("read email error: %s\n", err.Error())
			continue
		}
		// "=?GBK?B?" + base64("每日信用管家", "GBK") + "?="
		if !strings.HasPrefix(mr.Header["Subject"][0], "=?gb2312?B?w7/I1dDF08O53LzS") {
			continue
		}
		// mail date
		md, _ := time.Parse("Mon, 2 Jan 2006 15:04:05 -0700", mr.Header.Get("Date"))
		log.Printf("parsing mail @ %s\n", md.Format("2006-01-02"))
		// parse body
		mediaType, params, _ := mime.ParseMediaType(mr.Header.Get("Content-Type"))
		if strings.HasPrefix(mediaType, "multipart/") {
			br := multipart.NewReader(mr.Body, params["boundary"])
			for {
				p, err := br.NextPart()
				if err == io.EOF {
					break
				}
				if err != nil {
					log.Printf("email body parse error: %s\n", err.Error())
					break
				}

				slurp, err := ioutil.ReadAll(p)
				if err != nil {
					log.Printf("email body parse error: %s\n", err.Error())
					break
				}
				if p.Header.Get("Content-Type") == "text/html; charset=\"gb2312\"" {
					html, err := base64.StdEncoding.DecodeString(string(slurp))
					if err != nil {
						log.Printf("decode body to html error: %s\n", err.Error())
						break
					}
					html, _ = simplifiedchinese.GBK.NewDecoder().Bytes(html)
					// parse html
					doc, err := goquery.NewDocumentFromReader(bytes.NewReader(html))
					if err != nil {
						log.Printf("parse html error: %s\n", err.Error())
						break
					}
					text := strings.TrimSpace(doc.Text())
					if md.Before(date20210526) {
						orders = append(orders, parseBefore20210526(text)...)
					} else {
						orders = append(orders, parse(doc)...)
					}
				}

			}
		}
	}

	if err := <-done; err != nil {
		return orders, err
	}
	return orders, nil
}

func fileParseCmb(file string) ([]fileBillOrder, error) {
	var lines = make([][]string, 0)
	f, r, err := pdf.Open(file)
	if err != nil {
		return nil, fmt.Errorf("fail to open file: %s", file)
	}
	defer func() {
		_ = f.Close()
	}()

	var year int
	var month int
	dateRe := regexp.MustCompile(`(\d{4})年(\d{2})月\d{2}日`)
	blankRe := regexp.MustCompile(`\s+`)
	// the x coordinate
	var x float64
	// the y coordinate
	var y float64
	var line string
	pages := r.NumPage()
	for i := 1; i <= pages; i++ {
		p := r.Page(i)
		texts := p.Content().Text
		// a text is just a word
		for _, text := range texts {
			// determine new line
			if math.Abs(text.Y-y) > 1 {
				// find year
				if year == 0 && dateRe.FindString(line) != "" {
					year, _ = strconv.Atoi(dateRe.FindAllStringSubmatch(line, -1)[0][1])
					month, _ = strconv.Atoi(dateRe.FindAllStringSubmatch(line, -1)[0][2])
				}
				// stop if reach end of file
				if strings.TrimSpace(line) == "本期账单金额" {
					break
				}
				if len(strings.Split(line, "|")) == 6 {
					// 1,100 to 1100
					line = strings.ReplaceAll(line, ",", "")
					// replace blank
					line = blankRe.ReplaceAllString(line, "")
					lines = append(lines, strings.Split(line, "|"))
				}
				x, y, line = 0, text.Y, ""
			}
			// determine split in one line
			if x != 0 && math.Abs(text.X-x) > 10 && len(strings.TrimSpace(line)) > 0 {
				line += "|"
			}
			x = text.X
			line += text.S
		}
	}
	// format data
	// determine if date cross year (transaction date increase in order)
	// skip title
	firstMon, _ := strconv.Atoi(strings.Split(lines[1][0], "/")[0])
	lastMon, _ := strconv.Atoi(strings.Split(lines[len(lines)-1][0], "/")[0])
	crossYear := (firstMon - lastMon) > 0

	// bill date
	billDate, _ := time.Parse("2006-01", fmt.Sprintf("%d-%02d", year, month))
	isBeforeDate20211031 := billDate.Before(date20211031)

	var orders []fileBillOrder

	var max int
	for _, l := range lines {
		// skip title
		if l[0] == "交易日" || l[0] == "SOLD" {
			continue
		}

		if isBeforeDate20211031 {
			// date format
			mon, _ := strconv.Atoi(strings.Split(l[0], "/")[0])
			if crossYear && mon >= max {
				max = mon
				l[0] = strconv.Itoa(year-1) + "-" + strings.ReplaceAll(l[0], "/", "-")
				l[4] = strconv.Itoa(year-1) + "-" + strings.ReplaceAll(l[4], "/", "-")
			} else {
				l[0] = strconv.Itoa(year) + "-" + strings.ReplaceAll(l[0], "/", "-")
				l[4] = strconv.Itoa(year) + "-" + strings.ReplaceAll(l[4], "/", "-")
			}

			orders = append(orders, fileBillOrder{
				TradingDate:   l[0],
				Name:          l[1],
				Amount:        l[2],
				TailNumber:    l[3],
				CreditDate:    l[4],
				TradingAmount: l[5],
			})
		} else {
			// date format
			mon, _ := strconv.Atoi(strings.Split(l[0], "/")[0])
			if crossYear && mon >= max {
				max = mon
				l[0] = strconv.Itoa(year-1) + "-" + strings.ReplaceAll(l[0], "/", "-")
				l[1] = strconv.Itoa(year-1) + "-" + strings.ReplaceAll(l[1], "/", "-")
			} else {
				l[0] = strconv.Itoa(year) + "-" + strings.ReplaceAll(l[0], "/", "-")
				l[1] = strconv.Itoa(year) + "-" + strings.ReplaceAll(l[1], "/", "-")
			}

			orders = append(orders, fileBillOrder{
				TradingDate:   l[0],
				CreditDate:    l[1],
				Name:          l[2],
				Amount:        l[3],
				TailNumber:    l[4],
				TradingAmount: l[5],
			})
		}
	}

	return orders, nil
}
